EZP-25403: Change JS Rest Client to use X-HTTP-Method-Override for be…

andrerom opened the pull request:
…tter http server/proxy/scanner compat

Issue: https://jira.ez.no/browse/EZP-25403

One alternative to fixing this, other is changing it all over where  `PUBLISH/PATCH/MOVE/COPY/SWAP` is used.

- [x] build
- [x] tests

dpobel commented:
I don't think it's the role of the ConnectionManager to do that. I would rather do it in the XmlHttpRequestConnection. That way, if someone really wants to use custom verbs, it's possible to use another Connection implementation. While at it, I would also deprecate MicrosoftXmlHttpRequestConnection to not have to do it here as well.

dpobel commented on src/ConnectionManager.js:
```diff
@@ -69,6 +69,14 @@ define(["structures/Response", "structures/Request", "structures/CAPIError"],
 
+        // All other method types then defined in standard is sent as header to avoid issues
+        // with proxies, http security tools and limited web servers like the one in PHP.
+        var standardMethods = ["OPTIONS", "GET", "HEAD", "POST", "PUT", "DELETE", "TRACE"];
```
nitpick: I would rather use an Object like in https://github.com/ezsystems/ez-js-rest-client/blob/master/src/authAgents/SessionAuthAgent.js#L88

andrerom replied:
Won't I have to deal with `x.hasOwnProperty(y)` then?

andrerom commented:
Moved to XmlHttpRequestConnection in f64da5d.

bdunogier commented:
> Also tested and working with php internal server /cc @bdunogier

Awesome. Not sure what else I can say, as long as @dpobel agrees :-)

bdunogier commented on src/connections/XmlHttpRequestConnection.js:
```diff
@@ -46,10 +48,20 @@ define(["structures/Response", "structures/CAPIError"], function (Response, CAPI
             callback(false, response);
         };
 
+        // All other method types then defined in standard is sent as header to avoid issues
```
I'd simplify this. What about 

 ```
Sent non-standard HTTP methods as headers.
Avoids problems with proxies, HTTP security tools and limited web servers.
```

bdunogier commented:
And +1, cool one.

dpobel commented:
unit test please :p

andrerom commented:
> Related to #72 (comment)

fixed in 3a8c009

> I'd simplify this. What about

also fixed in 3a8c009

> unit tests please :p

Then I suggest someone else takes over for that and generating the combined files, I don't have setup for this anytime soon.

andrerom commented:
ready for review with all changes now @dpobel @yannickroger @StephaneDiot

dpobel commented on src/connections/MicrosoftXmlHttpRequestConnection.js:
```diff
@@ -9,6 +9,7 @@ define(["structures/Response", "structures/CAPIError"], function (Response, CAPI
      *
      * @class MicrosoftXmlHttpRequestConnection
      * @constructor
+     * @deprecated
```
maybe `@deprecated since 1.2`. I would also add in a warning when this object instantiated (similar to https://github.com/ezsystems/ez-js-rest-client/pull/71/files#diff-1cddad75741914589caf23a4815ed7abR1421)

andrerom replied:
fixed

dpobel commented on test/XmlHttpRequestConnection.tests.js:
```diff
@@ -139,6 +139,40 @@ define(function (require) {
+                    mockCallback
+                );
+
+                expect(mockXMLHttpRequest.prototype.open).toHaveBeenCalled();
```
I know the others tests are not written like that... but this can be simplified a bit with:

```js
expect(mockXMLHttpRequest.prototype.open).toHaveBeenCalledWith("POST", "/", true, "login", "password");                expect(mockXMLHttpRequest.prototype.setRequestHeader).toHaveBeenCalledWith("testHeader", "testHeaderValue");
expect(mockXMLHttpRequest.prototype.setRequestHeader).toHaveBeenCalledWith("X-HTTP-Method-Override", "PUBLISH");                expect(mockXMLHttpRequest.prototype.setRequestHeader.calls.length).toEqual(2);
expect(mockXMLHttpRequest.prototype.send).toHaveBeenCalledWith({testBody: ""});
```

dpobel commented on test/XmlHttpRequestConnection.tests.js:
```diff
@@ -139,6 +139,40 @@ define(function (require) {
+                expect(mockCallback.mostRecentCall.args[1]).toEqual(jasmine.any(Response)); // response
+
+                // cleanup
+                mockRequest.method = "GET";
```
I guess the initialization of `mockRequest` should be moved to `beforeEach` and this line can be removed.

andrerom replied:
yes, but then several things fail ;)

dpobel replied:
:( I can not tell I'm surprised. So if you can clean up, please do :)

andrerom commented:
@dpobel tests somewhat cleaned up, so hopefully a bit better now.

dpobel commented:
+1
small remark: usually we don't rebuild the library in the PR (to ease review mainly) and we do it after merge (at some point we should configure Travis to do that...)

andrerom commented:
> small remark: usually we don't rebuild the library in the PR (to ease review mainly) and we do it after merge (at some point we should configure Travis to do that...)

ok, will try to remember next time, makes sense to avoid conflicts between PR's as well :)
